Refine access note descriptions in diagnostics

beccadax opened the pull request:
This PR will finalize the way diagnostics describe when they're caused by access notes.

Specifically, the pre-existing access note remarks are now only emitted for valid attributes; for invalid attributes, the diagnostic is augmented with text explaining that it was added for an access note and has been disabled:

```
class MyClass: NSObject {
    // Both of these methods have ObjC: true in access notes

    func good(_: Int) {}
    // remark: implicitly added '@objc' to this method, as specified by access note for <reason string>
    // note: add '@objc' explicitly to silence this warning

    func bad(_: Int?) {}
    // remark: ignored access note: the type of the parameter cannot be represented in Objective-C;
    //         did not implicitly add '@objc' to this method, even though it was specified by access
    //         note for <reason string>
}
```

To accomplish this behavior, this PR makes it possible for a diagnostic to take a `DiagnosticInfo *` as a parameter; it then adds an `InFlightDiagnostic::wrapIn()` method which uses this capability. Thus, an error like `diag::objc_invalid_on_func_single_param_type` can be wrapped in `diag::wrap_invalid_attr_added_by_access_note` to add verbiage relating to access notes.

This PR also:

* Emits one access note success remark covering all attributes added to each declaration, instead of separate ones for each attribute.
* Softens selector conflicts caused by access notes into remarks.
* Adds a command-line option which controls whether and how access note remarks are emitted; in particular, you can use it to strengthen ignored access notes back to errors.

~~I will probably be tweaking this further before I'm finished with this PR.~~

Fixes rdar://71872575.

beccadax committed:
Emit access note remarks only for valid attributes

This requires deferring emission until the end of typechecking. A future change will emit access-note-related notes when an attribute is invalidated.

beccadax committed:
Mention access note when describing ObjCReason

beccadax committed:
Describe ObjC reason for effectful properites

beccadax committed:
Add option to control access note diagnostics

This will allow teams writing access notes to use -Raccess-note=all-validate to check that their access notes are correct, or teams working around problems to use -Raccess-note=failures or -Raccess-note=none to suppress diagnostics.

beccadax committed:
Don’t emit success remarks for bad @objc names

beccadax committed:
Rephrase all access note remarks

• There is now one access note success remark and fix-it per declaration, not per attribute/modifier.
• Failure remarks have been rephrased to better emphasize the cause of the failure.
• The wording of other access note remarks and notes have been changed to follow a similar formula.

beccadax committed:
[NFC] Correct selector conflict sorting rules

When two members have the same ObjC selector, there’s a test used to make sure that we’re “giving” the selector to the more “authoritative” of the two. However, its logic is not symmetrical, which I suspect could cause misbehavior in some edge cases. This change formalizes the logic in a way that eliminates the asymmetry.

beccadax committed:
Soften ObjC selector conflicts from access notes

beccadax committed:
“Wrap” invalid access note diagnostics

This shows up better in Xcode, which unfortunately doesn’t really display notes very clearly.

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci please test

xymus commented on include/swift/AST/DiagnosticEngine.h:
```diff
@@ -515,6 +528,42 @@ namespace swift {
+    ///     diagnostic currently in \c *this.
+    /// \li The location, ranges, decl, fix-its, and behavior limit of the
+    ///     diagnostic currently in \c *this.
+    InFlightDiagnostic &wrapIn(const Diagnostic &wrapper);
```
Can you tell me more about the advantage of wrapping this diagnostic in the Xcode display? Should we apply the same trick to more kind of diagnostics or is this specific to access notes?

beccadax replied:
Basically, Xcode is not very good at displaying notes in its UI. Wrapping a diagnostic produces a longer message, but it's a *single* message that will actually be displayed.

I've been thinking that it might be good to turn `diag::module_interface_build_failed` and friends into wrapping diagnostics around the underlying errors, since the underlying errors often seem to go unnoticed by users, but are usually a better way to resolve the problem. But that's a different pull request.

xymus reviewed (approved):
This looks good! I'll be curious to see a demo of it with all of the new diagnostics refinements.

beccadax committed:
Appease tyrannical Python linter

beccadax committed:
Share diagnoseAndRemoveAttr()

Merge together several helpers and code patterns for “diagnose/fix-it/invalidate bad attribute” into helper functions in TypeChecker.h.

This requires minor test changes in some places where we’re testing ObjC interop without importing Foundation; it’s otherwise NFC.

beccadax commented:
@swift-ci please test

beccadax commented:
@swift-ci smoke test and merge

beccadax commented:
@swift-ci smoke test and merge
